crates/vcz_http_server/src/lib.rs

```diff
@@ -0,0 +1,654 @@
+    NoInfoHash,
+
+    #[error("invalid info hash in the URI path")]
+    InvalidInfoHash,
+
+    #[error("a URI path component is not encoded with UTF-8")]
+    FromUtf8,
+}
+
+/// Web routing.
+/// Validates and converts a URI path to a file in a torrent.
+/// If it returns `Ok((info_hash, path))`, then `uri_path` refers
+/// to the file in the torrent identified by `info_hash`
+/// at `path` relative to the root of the torrent.
+fn parse_uri_path(
+    uri_path: &str,
+) -> Result<(InfoHash, Vec<String>), ParseUriPathError> {
+    let mut uri_path = uri_path.split('/').collect::<VecDeque<_>>();
+    match uri_path.pop_front() {
+        None => Err(ParseUriPathError::Split),
+        Some(first) => if !first.is_empty() {
+            Err(ParseUriPathError::StartSlash)
+        } else {
+            match uri_path.pop_front() {
+                None => Err(ParseUriPathError::NoInfoHash),
+                Some(info_hash) => match decode_info_hash(info_hash) {
+                    Err(_) => Err(ParseUriPathError::InvalidInfoHash),
+                    Ok(info_hash) => {
+                        let path: Result<Vec<_>, _> = Result::from_iter(
+                            uri_path
+                                .into_iter()
+                                .map(|s| urlencoding::decode(s)
+                                    .map(std::borrow::Cow::into_owned)
+                                )
+                        );
+                        match path {
+                            Err(_) => Err(ParseUriPathError::FromUtf8),
+                            Ok(path) => Ok((info_hash, path)),
+                        }
+                    },
+                },
+            }
+        },
+    }
+}
+
+
+
+#[derive(Debug, thiserror::Error)]
+pub enum CallError<A, E> {
+    #[error("error when sending an argument: {0}")]
+    Sender(mpsc::error::SendError<A>),
+
+    #[error("error when receiving a result: {0}")]
+    Receiver(oneshot::error::RecvError),
+
+    #[error("error returned by the callee: {0}")]
+    Callee(E),
+}
+
+/// Makes a call through synchronization channels.
+/// Sends `arg(resp_sender)` to `arg_sender` and receives a result
+/// from `resp_receiver` where `resp_sender` and `resp_receiver`
+/// are parts of the same oneshot channel.
+pub async fn channel_call<A, R, E>(
+    arg_sender: mpsc::Sender<A>,
+    arg: impl FnOnce(oneshot::Sender<Result<R, E>>) -> A,
+) -> Result<R, CallError<A, E>> {
+    let (resp_tx, resp_rx) = oneshot::channel();
+    let () = arg_sender.send(arg(resp_tx)).await.map_err(CallError::Sender)?;
+    resp_rx.await.map_err(CallError::Receiver)?.map_err(CallError::Callee)
+}
+
+
+
+type CallDiskError = CallError<DiskMsg, SliceError>;
+
+type ReadOutput = Result<Bytes, CallDiskError>;
+
+/// The purpose of this trait is to give a name
+/// to the return type of [`SliceReader::read`] (future).
+/// The name is `<SliceReader as SliceReaderI>::ReadFuture`.
+trait SliceReaderI {
```

Comment:
```suggestion
trait ReadSlice {
```

1. I think you meant `I` -> `Interface`. It's wrong because Rust doesn't have interfaces, and traits work differently from interfaces. We don't add the `I` postfix to trait names in Rust.
2. I suggest you rename it to `ReadSlice`, like [`AsyncRead`](https://docs.rs/tokio/latest/tokio/io/trait.AsyncRead.html), [`ReadBytesExt`](https://docs.rs/byteorder/latest/byteorder/trait.ReadBytesExt.html), etc.

Reply:
Transforming names makes it hard for me to find them later.

There is no better alternative for “I”. At least, “I” bears some intuition.

Reply:
> Transforming names makes it hard for me to find them later.

What do you mean "to find"?

> There is no better alternative for “I”. At least, “I” bears some intuition.

You don't need the alternative to "I". You should change the trait name completely :upside_down_face: (but I'm not insist and already approved the PR)